modules/foreman_domain.py

```diff
@@ -0,0 +1,187 @@
+    description: foreman user password
+    required: true
+  verify_ssl:
+    description: verify ssl connection when communicating with foreman
+    default: true
+    type: bool
+  state:
+    description: domain presence
+    default: present
+    choices: ["present", "absent"]
+'''
+
+EXAMPLES = '''
+- name: domain
+  foreman_domain:
+    name: "example.org"
+    description: "Example Domain"
+    locations:
+      - "Munich"
+    organizations:
+      - "ATIX"
+    server_url: "https://foreman.example.com"
+    username: "admin"
+    password: "secret"
+    verify_ssl: False
+    state: present
+'''
+
+RETURN = ''' # '''
+
+try:
+    from ansible.module_utils.ansible_nailgun_cement import (
+        create_server,
+        ping_server,
+        find_domain,
+        find_location,
+        find_organization,
+        find_smart_proxy,
+        naildown_entity_state,
+        sanitize_entity_dict,
+    )
+
+    from nailgun.entities import (
+        Domain,
+    )
+
+    HAS_NAILGUN_PACKAGE = True
+except:
+    HAS_NAILGUN_PACKAGE = False
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.foreman_helper import handle_no_nailgun
+
+
+# This is the only true source for names (and conversions thereof)
+name_map = {
+    'name': 'name',
+    'description': 'fullname',
+    'dns_proxy': 'dns',
+    'locations': 'location',
+    'organizations': 'organization',
+}
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            server_url=dict(required=True),
+            username=dict(required=True),
+            password=dict(required=True, no_log=True),
+            verify_ssl=dict(type='bool', default=True),
+            name=dict(required=True),
+            description=dict(),
+            dns_proxy=dict(),
+            locations=dict(type='list'),
+            organizations=dict(type='list'),
+            state=dict(choices=['present', 'absent'], default='present'),
+        ),
+        supports_check_mode=True,
+    )
+
+    handle_no_nailgun(module, HAS_NAILGUN_PACKAGE)
+
+    domain_dict = dict(
+        [(k, v) for (k, v) in module.params.items() if v is not None])
+
+    server_url = domain_dict.pop('server_url')
+    username = domain_dict.pop('username')
+    password = domain_dict.pop('password')
+    verify_ssl = domain_dict.pop('verify_ssl')
+    state = domain_dict.pop('state')
+
+    try:
+        create_server(server_url, (username, password), verify_ssl)
+    except Exception as e:
+        module.fail_json(msg="Failed to connect to Foreman server: %s " % e)
+
+    ping_server(module)
+    try:
+        # Try to find the Domain to work on
+        entity = find_domain(module, name=domain_dict['name'], failsafe=True)
+    except Exception as e:
+        module.fail_json(msg='Failed to find entity: %s ' % e)
+
+    if 'dns_proxy' in domain_dict:
+        domain_dict['dns_proxy'] = find_smart_proxy(module, domain_dict['dns_proxy'])
+
+    if 'locations' in domain_dict:
```

Comment:
You could implement `find_locations` in ansible_nailgun_cement like [this](https://github.com/theforeman/foreman-ansible-modules/blob/master/module_utils/ansible_nailgun_cement.py#L282), or use [`find_entities`](https://github.com/theforeman/foreman-ansible-modules/blob/master/module_utils/ansible_nailgun_cement.py#L157). Same for organizations.

Reply:
Agreed, I added `find_locations` and `find_organizations` similar to `find_repositories`